src/ql/com/dec/unitary.cc

```diff
@@ -665,6 +624,88 @@ static Int recursiveRelationsForUnitaryDecomposition(
     }
 }
 
+/**
+ * Does state preparation, results in a circuit for which A|0> = |psi> for given state psi, stored as the array in Unitary
+ */
+ir::compat::GateRefs Unitary::prepare_state(const utils::Vec<utils::UInt> &qubits){
+    UInt nqubits = qubits.size();
+    if (((UInt) 1) << nqubits != array.size()){
+        QL_FATAL("Length of state prepatation vector does not match number of qubits! Expected vector of size " << (1 << nqubits) << " but got vector of size " << array.size());
+    }
+    ir::compat::GateRefs c;
+    Vec<Complex> statevector = array;
+    int n = statevector.size() / 2;
+    Vec<Real> phi;
+    Vec<Real> theta;
+    UnitaryDecomposer decomposer;
+
+    QL_DOUT("Preparing state with state vector " << statevector);
+    while (n > 0)
+    {
+        for (int c = 0; c < n; c++)
+        {
+            Complex A = statevector[2 * c];
+            Complex B = statevector[2 * c + 1];
+            Vec<Real> abg = decomposer.zyz_decomp(A, B, 1);
+            theta.push_back(-abg[1]); // beta
+            phi.push_back(abg[0]);  // gamma
+            statevector[c] = cos(0.5 * theta.back()) * exp(Complex(0, -0.5) * phi.back()) * A - sin(0.5 * theta.back()) * exp(Complex(0, 0.5) * phi.back()) * B;
+        }
+        statevector.resize(n);
+        QL_DOUT("New statevector: " << statevector);
+        n = n >> 1; // n = n/2
+    }
+    QL_DOUT("Qubits: " << qubits << " qubits.size(): " << qubits.size() << " phi: " << phi << " theta: " << theta);
+    c.emplace<ir::compat::gate_types::RY>(qubits.back(), theta.back());
+    c.emplace<ir::compat::gate_types::RZ>(qubits.back(), phi.back());
+    if (nqubits > 1){
+        UInt start_i = phi.size() - 3;
+        UInt ngates;
+        decomposer.genMk(qubits.size());
+        Eigen::VectorXd temp;
+        Eigen::VectorXd tr;
+        UInt idx;
+        for (UInt i = 1; i < nqubits; i++)
+        {
+            ngates = 1 << i;
+            QL_DOUT("Sending indices " << start_i << " until " << (start_i + ngates) << " to multicontrolled z and y. i=" << i);
+            Eigen::CompleteOrthogonalDecomposition<Eigen::MatrixXd> dec(decomposer.genMk_lookuptable[i - 1]);
+            temp = Eigen::Map<Eigen::VectorXd>(theta.data() + start_i, ngates);
+            tr = dec.solve(temp);
+
+            // The first one is always controlled from the next qubit to the last qubit
+            c.emplace<ir::compat::gate_types::RY>(qubits[nqubits-i-1], tr[0]);
+            c.emplace<ir::compat::gate_types::CNot>(qubits[nqubits-i], qubits[nqubits-i-1]);
+            for (UInt j = 1; j < (ngates - 1); j++){
+                idx = i - log2(((j)^((j)>>1))^((j+1)^((j+1)>>1)));
```

Comment:
It would be nice to extract this kind of computation as an auxiliary function.
- This gives an explicit name to the piece of code, which gives information to the reader
- It makes the code clearer (generally speaking, any function longer than 20 lines should be avoided, IMO)
- It makes it possible to write unit tests for such pieces of code.

More on unit tests:
- They validate the correctness of small pieces of code.
- They allow to check for corner cases easily.
- They provide documentation to the reader, as they showcase what the functions do.
- They provide guarantee that one is not changing the behavior of the code when refactoring.

Test coverage is important to write correct software.